Generic H-Bridge driver + H-Bridge switch implementation

FaBjE opened the pull request:
feature/hbridge
--------------------

# What does this implement/fix? 

This pull request combines several (partial) h-bridge driver implementations into one generic h-bridge driver.
I added DC motor transition/breaking implementation to this hbridge driver.
Currently it is a reusable software class and not usable as a instance to end users but this could be added.

I ported the H-Bridge fan and H-Bridge light component to the new generic H-bridge driver implementation for code-reuse.

I also added a "H-Bridge switch" component. This component can abstract linear actuators / motorized valve controllers as a switch. (See documentation pull request for more info)
This greatly simplifies using them in home-assistant.

All changes are backwards compatible.

I've claimed ownership of the H-Bridge fan and H-Bridge light component, I changed them a fair bit. 
If anyone wants it back, let me know.

## Types of changes

- [ ] Bugfix
- [X] New feature (Add H-Bridge switch + generic HBridge driver)
- [ ] Breaking change
- [ ] Other

**Related issue or feature (if applicable):**

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** 
esphome/esphome-docs#1835


## Test Environment

- [ ] ESP32
- [ ] ESP32 IDF
- [x] ESP8266

(I have no ESP32 available sorry, I wrote the code with ESP32 compatibility in mind. I expect no issues as nothing directly uses low level hw functions)

## Example entry for `config.yaml`:

```yaml

# Output list
output:
  - platform: esp8266_pwm
    id: output_fan_a
    pin: D1
    frequency: 50 Hz
  - platform: esp8266_pwm
    id: output_fan_b
    pin: D2
    frequency: 50 Hz
  - platform: esp8266_pwm
    id: output_fan_enable
    pin: D3
  - platform: gpio
    id: output_fan_oscillating
    pin: D4
  - platform: esp8266_pwm
    id: output_light_a
    pin: D5
    frequency: 100 Hz
  - platform: esp8266_pwm
    id: output_light_b
    pin: D6
    frequency: 100 Hz
  - platform: esp8266_pwm
    id: output_switch_a
    pin: D7
    frequency: 50 Hz
  - platform: esp8266_pwm
    id: output_switch_b
    pin: D8
    frequency: 50 Hz
    

# Fan test config
fan:
  - platform: hbridge
    id: hbridge_fan_configid
    name: "H-Bridge Fan"
    pin_a: output_fan_a
    pin_b: output_fan_b
    enable_pin: output_fan_enable
    decay_mode: FAST
    oscillation_output: output_fan_oscillating
    min_speed: 0.05
    brake_when_stopped: 2000ms
    brake_buildup_when_stopped: 3000ms
    acceleration: 8000ms
    deceleration: 4000ms

# Light test config
light:
  - platform: hbridge
    id: hbridge_light_configid
    name: "H-Bridge Light"
    pin_a: output_light_a
    pin_b: output_light_b


# Switch test config    
switch:
  - platform: hbridge
    id: hbridge_switch_configid
    name: "H-Bridge switch"
    pin_a: output_switch_a
    pin_b: output_switch_b
    duration: 10000ms

```

## Checklist:
  - [X] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).
  
If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

FaBjE committed:
Add generic HBridge driver + Update HBridge-Fan and HBridge-LightOutput

BREAKING: Removed "decay_mode" See: #2871

FaBjE committed:
Add hbridge_switch: Voltage-reversing switch implementation

FaBjE committed:
Generalize HBridge config parsing + Add Transitions

Set default config to no transition

oxan commented on esphome/components/hbridge/__init__.py:
```diff
@@ -1,3 +1,62 @@
+from esphome.components import output
+from esphome.const import CONF_PIN_A, CONF_PIN_B, CONF_ENABLE_PIN
 
 hbridge_ns = cg.esphome_ns.namespace("hbridge")
```
The new `HBridge` parent class should be properly reflected in the code generation phase:
* Define the class here using `HBridge = hbridge_ns.class_("HBridge")`
* Add an ID of that type to the config schema: `cv.GenerateID(): cv.declare_id(HBridge)`
* All child classes should also inherit from the HBridge class: `HBridgeFan = hbridge_ns.class_("HBridgeFan", fan.FanState, hbridge.HBridge)`

This allows the config validation phase to check whether IDs reference the correct types and objects.

FaBjE replied:
I'm trying to implement this, but whenever i put `cv.GenerateID(): cv.declare_id(HBridge)` in the HBridge base class it runs into problems. I either get errors that components are "Not registered to inherit from Component or registered twice" or, the generated C++ main code sees every instance as C++ HBridge type and tries to create that as an instance. which fails because it is not a child-class type.

I've put `cv.GenerateID(): cv.declare_id(xxxx)` in every child-class config and that works fine.

Is this what you meant?

oxan replied:
> I either get errors that components are "Not registered to inherit from Component or registered twice"

Ah yes, if we also change `HBridge` to inherit from `Component`, that should be done in the class definition here as well:

```python
HBridge = hbridge_ns.class_("HBridge", cg.Component)
```

FaBjE replied:
I removed the line as I was unable to get it working with it.

oxan commented on esphome/components/hbridge/__init__.py:
```diff
@@ -1,3 +1,62 @@
+)
+
+
+async def hbridge_config_to_code(config, var):
```
The usual convention is:
```suggestion
async def setup_hbridge(var, config):
```

FaBjE replied:
I used a different function name on purpose as the hbridge class cannot be used directly and this was more of an "helper function" (I will change it)

oxan replied:
Yes, that's true, but e.g. the `Climate` or `Fan` class also can't be used on it's own, and they do follow the `setup_` pattern. It's not really important, just a nice bit of consistency :)

oxan commented on esphome/components/hbridge/hbridge.h:
```diff
@@ -0,0 +1,85 @@
+  HBRIDGE_TRANSITIONING_STATE_SHORTING_BUILDUP = 1,
+  HBRIDGE_TRANSITIONING_STATE_FULL_SHORT = 2,
+  HBRIDGE_TRANSITIONING_STATE_DUTYCYCLE_TRANSITIONING = 3,
+};
```
Can we change this to use scoped enumerations? We can also drop the long prefix then.

oxan commented on esphome/components/hbridge/hbridge.h:
```diff
@@ -0,0 +1,85 @@
+
+  // ESPHome component interfacing
+  void setup();
+  void loop();
```
You need to inherit from `Component` for these methods to be usable.

oxan commented on esphome/components/hbridge/hbridge.cpp:
```diff
@@ -0,0 +1,359 @@
+    uint32_t mode_duration = millis() - this->transition_mode_start_time_;
+    uint32_t ms_since_last_step = millis() - this->transition_last_step_time_;
+
+    // ESP_LOGD(TAG, "TS:%d dT:%d [HM:%d|%f]", this->transition_state_, ms_since_last_step, this->current_mode_,
+    // this->current_relative_dutycycle_);
```
We don't want commented code. If this is still useful, it should be logged at e.g. the very verbose level (`ESP_LOGVV`), otherwise it should be dropped.

oxan commented on esphome/components/hbridge/hbridge.cpp:
```diff
@@ -0,0 +1,359 @@
+
+void HBridge::transition_to_state(HBridgeMode target_mode, float target_dutycycle, float dutycycle_delta_per_ms,
+                                  uint32_t shorting_buildup_duration_ms = 0, uint32_t full_short_duration_ms = 0) {
+  // ESP_LOGD(TAG, "shorting_buildup_duration_ms: target_mode:%d target_dutycycle:%.2f dutycycle_delta_per_ms:%f ms
+  // shorting_buildup_duration_ms:%d ms full_short_duration_ms:%d ms", target_mode, target_dutycycle,
+  // dutycycle_delta_per_ms, shorting_buildup_duration_ms, full_short_duration_ms);
```
Same here.

oxan commented on esphome/components/hbridge/hbridge.cpp:
```diff
@@ -0,0 +1,359 @@
+  // Check argument range
+  if (relative_dutycycle > 1 || relative_dutycycle < -1) {
+    return;  // Invalid argument
+  }
```
Maybe clamp it to (-1, 1) instead? 
```c++
relative_dutycycle = clamp(relative_dutycycle, -1.0f, 1.0f);
```

oxan commented on esphome/components/hbridge/light/hbridge_light_output.h:
```diff
@@ -24,34 +22,41 @@ class HBridgeLightOutput : public PollingComponent, public light::LightOutput {
+      HBridge::set_output_state_(
+          HBRIDGE_MODE_DIRECTION_B,
+          this->light_direction_b_duty_);  // Use protected function to prevent a flood of debug prints
+      this->direction_a_update_ = false;
```
Can you just merge this into `loop()` and drop this method and the `PollingComponent` parent class? it gets called with the same frequency anyway.

FaBjE replied:
I will do that, the original component was build this way. I assumed it was for a reason but it seemed a bit odd to me as well.

oxan replied:
Yes, this isn't related to your changes, but a nice bit of cleanup while we're at it :)

oxan commented on esphome/components/hbridge/switch/hbridge_switch.cpp:
```diff
@@ -0,0 +1,110 @@
+static const char *const TAG = "switch.hbridge";
+
+void HBridgeSwitch::setup() {
+  ESP_LOGCONFIG(TAG, "Setting up Valve Actuator: '%s'...", this->name_.c_str());
```
```suggestion
  ESP_LOGCONFIG(TAG, "Setting up H-Bridge Switch: '%s'...", this->name_.c_str());
```

oxan commented on esphome/components/hbridge/switch/hbridge_switch.cpp:
```diff
@@ -0,0 +1,110 @@
+    this->turn_off();
+}
+
+void HBridgeSwitch::loop() { HBridge::loop(); }
```
If you make `HBridge` inherit from `Component`, this can be dropped.

FaBjE replied:
I was not sure to let HBridge itself inherit from component or the child classes. As HBridge is more of an abstract class at the moment.

The function here is pretty useless I agree, but HBridgeFan and HBridgeLight both use the loop function for their own functionality.

If I were to override that I would override/remove the function from HBridge (base) .
I would need to add a call to the base/HBridge class as well. Which brings me pretty much back to the state it is now.

Should we leave it this way or do the override + base class call method?

oxan replied:
My view here is that the `HBridge` class requires `loop()` to be called, so child classes must inherit from `Component`. Why force all users to do it manually, instead of doing it in `HBridge` already and removing that burden from child classes that don't need it?

You're right that for the fan/light it doesn't make any difference as they override it themselves as well. I don't think that's an issue.

FaBjE replied:
Agreed. I'll change it.

FaBjE replied:
I'm trying to implement this but I run into all kinds of ambiguity errors.

For example, i have the `HBridge ` class inherit from `Component` now, but I get ambiguity errors in HBridgeFan because `fan::FanState` also inherits from `Component`
I can't think of anything how to solve this than letting every child class inherit from `Component` when needed.
Any suggestions?

Edit:
This is essentially the "Diamond problem" https://www.makeuseof.com/what-is-diamond-problem-in-cpp/
We could fix it using virtual inheritance but that would require (a lot of) changes to the core of esphome.

oxan replied:
Ah, yeah, this is a problem because the fan component was a bit of a mess; entities such as `FanState` shouldn't inherit from `Component`. That has been fixed in #2848, which is now merged into dev, but unfortunately also changed the H-Bridge fan.

The best option is probably if you could merge `dev` into your branch first.

oxan reviewed (commented):
Thanks! This looks like a good improvement. I didn't check the transition logic, but I'll assume you tested it :)

Some comments inline:

FaBjE commented:
Thank you for your extensive and valid feedback! I have second to none experience in Python so it is very much appreciated.

I will process your points and update this branch. I've added some comments per-point to clarify.

FaBjE committed:
Process pull-request feedback

https://github.com/esphome/esphome/pull/3086
Fix clang-tidy errors
Update codeowners
Run clang formatter on c/cpp files
Fix python lint errors
Run black formatter on python files
Add newlines for Lint check

FaBjE committed:
Remove line that causes all configs to be HBridge instead of child type components

FaBjE commented:
I rebased my feature on the dev branch and did a bit of a cleanup.
I processed all your feedback points and verified the workings :)

It is all working now except for the `cv.GenerateID(): cv.declare_id(HBridge)` line. 
When I leave that in the sourcecode generated uses all `HBridge` types instead of `HBridgeSwitch` etc.
I removed it in the last commit https://github.com/esphome/esphome/pull/3086/commits/56a816cc0de8f526734d619f046d3905eb7f759b
Could you check if this is what you intended?


I changed the codeowner marker to myself as well. I changed so much that it seems the right thing to do.
If anyone wants it back no problem ;)

WeekendWarrior1 commented:
https://github.com/esphome/issues/issues/2871#issuecomment-1046149764

While I think that creating a generic h-bridge parent component is useful, the Decay mode issues mentioned in this issue are not issues at all, but working as intended, and definitely should not be removed/changed.

https://deploy-preview-1835--esphome.netlify.app/components/fan/hbridge.html#transitions-explained
> Fortunately you can electrically brake a motor by shorting its windings. Just as driving a car the slamming on the brakes is not very pleasant experience for your circuit.

Are there any citations available for this statement?

FaBjE commented:
Thank you for your reply and explanation @WeekendWarrior1 both here and in the issue I opened.
My goal is to provide a solid universal hbridge driver. As I didn't receive a response on my issue at the time I continued under the impression it was not supposed to be this way.

I've read the articles you linked and did some research of my own. 
And I'm at the point I'm not sure if we use different terminology for the same thing, or I'm misunderstanding something.
So let me ask you really blunt (trying to learn something here)

Is the decay mode supposed to be applied on each PWM cycle (when driving the motor at a steady speed)? or is the decay mode the strategy you apply when "not driving the motor" anymore?
I think you mean the first but I can't find any hard-statements for that in the articles I read.

> Are there any citations available for this statement?
No. Seeing this now I should phrase that differently.

WeekendWarrior1 commented:
> Is the decay mode supposed to be applied on each PWM cycle (when driving the motor at a steady speed)? or is the decay mode the strategy you apply when "not driving the motor" anymore?
I think you mean the first but I can't find any hard-statements for that in the articles I read.

Yes, the first.
Decay mode is specific to when a motor is being PWM speed controlled, and should occur every PWM cycle.

I would also suggest rethinking and simplifying your `transition_delta_per_ms`, `transition_short_buildup_duration` and `transition_full_short_duration` variables to simply `acceleration`/`deceleration` (possibly in the unit of `speed percentage/second`). I think the simplicity would be nice for end users.

Finally, from your docs:
>Trying to speed up the motor inertia will require to motor to use large amounts of current to get things going. You can reduce the current/load on your system by gradually increasing the speed.

I do not believe that gradually increasing the speed reduces the current required to get the motor spinning (and please also consider that a H-Bridge pin is really only ever fully on, or fully off, so there will always be current spikes).
Another complication is that depending on how small/short the pulses are at the start of acceleration (eg 1% speed),
![image](https://user-images.githubusercontent.com/5209758/154869098-656ab3dd-6569-4681-8dbc-e2503181534d.png)
The small pulses will be initially completely ineffective (on the graph above, up until 20% speed) and just cause unnecessary load and heat.

FaBjE committed:
Restore (current) decay mode functionality

FaBjE committed:
Give hbridge_light_output its own .cpp file to allow log output (and fix lint error)

FaBjE commented:
> Yes, the first.
>Decay mode is specific to when a motor is being PWM speed controlled, and should occur every PWM cycle.

I've restored/re-implemented the decay mode functionality.

> I would also suggest rethinking and simplifying your transition_delta_per_ms, transition_short_buildup_duration and transition_full_short_duration variables to simply acceleration/deceleration (possibly in the unit of speed percentage/second). I think the simplicity would be nice for end users.

I agree this is not the most userfriendly option. In a way the "delta_per_ms" already a percentage to a time limit.
But renaming/adding a "ac/de-celeration speed" parameter sounds good. Let me give it some thought.

> I do not believe that gradually increasing the speed reduces the current required to get the motor spinning (and please also consider that a H-Bridge pin is really only ever fully on, or fully off, so there will always be current spikes).

I removed the claim to be sure.

> Another complication is that depending on how small/short the pulses are at the start of acceleration (eg 1% speed),
> The small pulses will be initially completely ineffective (on the graph above, up until 20% speed) and just cause unnecessary load and heat.

True, but it depends on the circuit/motor where the "cutoff" is that they are not effective anymore.
So I'm hesitant to limit them.


> PS. Apologies for the delay on commenting on your Github issue as well. In theory I should get notifications when an issue is tagged on a component I codeown, but that wasn't the case.

Thats ok. I think the problem was the name "WeekendWarrior" instead of "WeekendWarrior1" was in the codeowners file.

FaBjE commented:
@oxan Should I do anything for you to be able to review? Or just wait patiently? 😉

jesserockz committed:
Merge branch 'dev' into feature/hbride_driver

jesserockz commented on esphome/components/hbridge/fan/__init__.py:
```diff
@@ -2,45 +2,29 @@
-# Actions
-BrakeAction = hbridge_ns.class_("BrakeAction", automation.Action)
-
+HBridgeFan = hbridge.hbridge_ns.class_("HBridgeFan", fan.FanState, hbridge.HBridge)
```
```suggestion
HBridgeFan = hbridge.hbridge_ns.class_("HBridgeFan", fan.Fan, hbridge.HBridge)
```

jesserockz commented on esphome/components/hbridge/switch/__init__.py:
```diff
@@ -0,0 +1,50 @@
+
+    cg.add(var.set_restore_mode(config[CONF_RESTORE_MODE]))
+
+    if CONF_DURATION in config:
+        cg.add(var.set_switching_duration(config[CONF_DURATION]))
```
Default exists so it is always in the config
```suggestion
    cg.add(var.set_switching_duration(config[CONF_DURATION]))
```

jesserockz commented on esphome/components/hbridge/__init__.py:
```diff
@@ -1,3 +1,80 @@
+        pin_enable = await cg.get_variable(config[CONF_ENABLE_PIN])
+        cg.add(var.set_hbridge_enable_pin(pin_enable))
+
+    if CONF_DECAY_MODE in config:
+        cg.add(var.set_hbridge_decay_mode(config[CONF_DECAY_MODE]))
+
+    # Transition settings
+    if CONF_TRANSITION_DELTA_PER_MS in config:
+        cg.add(
+            var.set_setting_transition_delta_per_ms(
+                config[CONF_TRANSITION_DELTA_PER_MS]
+            )
+        )
+
+    if CONF_TRANSITION_SHORT_BUILDUP_DURATION in config:
+        cg.add(
+            var.set_setting_transition_shorting_buildup_duration_ms(
+                config[CONF_TRANSITION_SHORT_BUILDUP_DURATION]
+            )
+        )
+
+    if CONF_TRANSITION_FULL_SHORT_DURATION in config:
+        cg.add(
+            var.set_setting_transition_full_short_duration_ms(
+                config[CONF_TRANSITION_FULL_SHORT_DURATION]
+            )
+        )
```
All of these have defaults
```suggestion
    cg.add(var.set_hbridge_decay_mode(config[CONF_DECAY_MODE]))

    # Transition settings
    cg.add(
        var.set_setting_transition_delta_per_ms(
            config[CONF_TRANSITION_DELTA_PER_MS]
        )
    )

    cg.add(
        var.set_setting_transition_shorting_buildup_duration_ms(
            config[CONF_TRANSITION_SHORT_BUILDUP_DURATION]
        )
    )

    cg.add(
        var.set_setting_transition_full_short_duration_ms(
            config[CONF_TRANSITION_FULL_SHORT_DURATION]
        )
    )
```

jesserockz reviewed (changes requested):
A few minor changes

FaBjE committed:
Apply suggestions from code review

Co-authored-by: Jesse Hills <3060199+jesserockz@users.noreply.github.com>

FaBjE commented:
Thank you for your review and valid points @jesserockz
I have changed them and verified the workings.

FaBjE committed:
Remove line of commented code

WeekendWarrior1 commented:
>>Another complication is that depending on how small/short the pulses are at the start of acceleration (eg 1% speed), The small pulses will be initially completely ineffective (on the graph above, up until 20% speed) and just cause unnecessary load and heat.

>True, but it depends on the circuit/motor where the "cutoff" is that they are not effective anymore.
So I'm hesitant to limit them.

If it's dependent on the circuit/motor, then it makes perfect sense to have it user configurable, eg. `min_speed`? Or `starting_speed`?
The reason why this is important is if someone has a slow acceleration (over 10s), but their motor is stalled until it's reached 50% speed, it's going to burn out the motor very fast (I have done this on a few TT DC plastic gearbox motors lol).


---
https://deploy-preview-1835--esphome.netlify.app/components/fan/hbridge.html#transition-configuration

All the transition configuration seems so needlessly complex and challenging to understand (I have plenty of experience with DC motors and I find it a little puzzling/confusing)

> Use the transition_delta_per_ms parameter for this. This will increment the speed with the set value every millisecond until the set speed has been reached.

> For example use 0.001 to stretch a 0 (still) to 1 (full speed) increment over 1 second.

There has to be a nicer way for the average end-user to configure this (and word it)
I still believe having it configurable as:
- `acceleration`: speed percentage/second (eg. )
or
- `acceleration`: seconds (eg. takes 1 second to ramp from 0%-100%)
Would be a massive improvement.

---
>- transition_short_buildup_duration (Optional, float): duration in milliseconds to buildup from no to full short. Defaults to 0 (no shorting/braking = coasting)

> - transition_full_short_duration (Optional, float): duration in milliseconds to apply full short (after optional buildup). Defaults to 0 (no shorting/braking = coasting)

Can you explain these a bit more?
Do they apply every PWM cycle?
Or just when using the brake action? (if yes, maybe they should just be an optional variable on the brake action?)
Is `transition_short_buildup_duration` some sort of deceleration setting that works the same as `transition_delta_per_ms`?
Is `transition_full_short_duration` just emulating how slow decay mode works?

WeekendWarrior1 commented:
Okay found your much better explanation in the code:
https://github.com/esphome/esphome/blob/2b71b76a04810d9b0eb238b311c43f840c0ae497/esphome/components/hbridge/hbridge.cpp#L128
```c++
  // --- Behaviour Examples:
  // Transition - Direction A 100% to Direction A 20% (No shorting is applied regardless of values):
  // |           A100% >> A20%            |
  // |< ---  transition_duration_ms  --- >|
  //
  // Transition - Direction A 100% to Direction B 100%, No shorting (duration set to 0):
  // | A100% >> A0% | Idle | B0% >> B100% |
  // |< ---   transition_duration_ms --- >|
  //
  // Transition - Direction A 100% to 0%, With shorting (duration set > 0):
  // | A0% |  Shorting 0% >> Shorting 100%  |       Shorting 100%      |
  //       |< shorting_buildup_duration_ms >|
  //                                        |< full_short_duration_ms >|
  //
  // Transition - Direction A 100% to Direction B 100%, With shorting (duration set > 0):
  // | A0% |  Shorting 0% >> Shorting 100%  |       Shorting 100%      |       B0% >> B100%       |
  //       |< shorting_buildup_duration_ms >|
  //                                        |< full_short_duration_ms >|
  //                                                                   |< transition_duration_ms >|

```
Okay, so basically the hbridge.brake action, but can automatically be applied at the end of travel.

So could the wording be more like:
- `brake_when_stopped`: optional, default 0s

eg. `brake_when_stopped`: .5s = 500ms of braking after hbridge has stopped

FaBjE commented:
@WeekendWarrior1
I agree on your proposal for the config options. I will start to work on implementing them.

FaBjE committed:
Hbridge: Add simpler config options (untested)

FaBjE committed:
Merge remote-tracking branch 'upstream/dev' into feature/hbride_driver

FaBjE committed:
HBridge Light: Reduce flickering when only one side enabled

FaBjE committed:
Hbridge: Fix mistakes

FaBjE committed:
HBridge: Fix clang errors

FaBjE commented:
I have changed the configuration parameters as @WeekendWarrior1 suggested. (with some minor tweaks)

-  `acceleration` (time) <- time from 0-100%
-  `deceleration` (time)<- time from 100-0%
-  `brake_when_stopped` (time)
-  `brake_buildup_when_stopped` (time) _For "Fast current decay" mode, ignored in "Slow current decay" mode_
-  `min_speed` (dutycycle) _Because this is more in-line with the [output component](https://esphome.io/components/output/index.html) "min_power" option_

And it is indeed much easier to config/use now :)
The code has been tested, and the documentation has been updated (https://github.com/esphome/esphome-docs/pull/1835)

See first post for the new example/test configuration.

I also had an idea to improve/reduce the flickering in the hbridge light implementation, I added that as well.

FaBjE commented:
The documentation pull request has automatically been marked as stale due to inactivity 😞 

I've put a lot of work into this. It would be unfortunate if it were all for nothing...

Please let me know what is keeping you from merging so I can fix it.

FaBjE committed:
Merge remote-tracking branch 'upstream/dev' into feature/hbride_driver
